third_party/move/move-binary-format/src/file_format.rs

```diff
@@ -2934,6 +2969,134 @@ pub enum Bytecode {
+        func param types match provided parameter types
+    "#]
+    #[gas_type_creation_tier_0 = "closure_ty"]
+    EarlyBindFunction(SignatureIndex, u8),
+
+    #[group = "closure"]
+    #[description = r#"
+        `InvokeFunction(|t1..tn|r with a)` calls a function value of the specified type,
+        with `n` argument values from the stack.
+
+        On top of the stack is the closure being evaluated, underneath the arguments:
+        `[c,vn,..,v1] + stack`. The type of the closure must match the type specified in
+        the instruction, with abilities `a` a subset of the abilities of the closure value.
+        A value `vi` on the stack must have type `ti`.
+
+        Notice that the type as part of this instruction is redundant for
+        execution semantics. Since the closure is expected to be on top of the stack,
+        it can decode the arguments underneath without type information.
+        However, the type is required for the current implementation of
+        static bytecode verification.
+
+        The arguments are consumed and pushed to the locals of the function.
+
+        Return values are pushed onto the stack from the first to the last and
+        available to the caller after returning from the callee.
+
+        During execution of the function conservative_invocation_mode is enabled;
+        afterwards, it is restored to its original state.
+    "#]
+    #[semantics = r#"
+        stack >> function_handle
+        let [func, k, [arg_0, .., arg_{k-1}]] = function_handle
+        // Information like the function signature are loaded from the file format
+        // and compared with the signature parameter.
+        n = func.num_params
+        ty_args = if func.is_generic then func.ty_args else []
+
+        n = func.num_params
+        stack >> arg_{n-1}
+        ..
+        stack >> arg_k
+
+        old_conservative_invocation_mode = conservative_invocation_mode
+        conservative_invocation_mode = true
+
+        let module = func.module
+        if module is on invocation_stack then
+           abort
+        else
+           invocation_stack << module
+
+        if func.is_native()
+            call_native(func.name, ty_args, args = [arg_0, .., arg_{n-1}])
+            current_frame.pc += 1
+        else
+            call_stack << current_frame
+
+            current_frame = new_frame_from_func(
+                func,
+                ty_args,
+                locals = [arg_0, .., arg_n-1, invalid, ..]
+                                           // ^ other locals
+            )
+
+        invocation_stack >> _
+        conservative_invocation_mode = old_conservative_invocation_mode
+    "#]
+    #[runtime_check_epilogue = r#"
+        // NOT: assert func visibility rules
+        for i in 0..#args:
+            ty_stack >> ty
+            assert ty == locals[#args -  i - 1]
+    "#]
+    #[gas_type_creation_tier_1 = "closure_ty"]
+    InvokeFunction(SignatureIndex),
```

Comment:
Why? `LdFunction` does not need to load until use? We store idx/name in a function value, and only when we call `Invoke` we resolve that index/name and load the function?

Reply:
The `SignatureIndex` here is a static type, and is needed for verifier to allow the stack depth checker to know how many parameters/results are being popped/pushed on the stack, since the stack depth analysis is independent of the type analysis in the verifier (if these 2 analyses were combined, then this need not be included in the instruction, as it should be easy to derive from context; I really don't know why there are bunch of independent passes in the verifier).

Execution will have to look at the function value (1) to get the early-bound parameters to pass, and (2) to get the FunctionHandle/FunctionInstantiation to (a) get its type to figure out the number and type of parameters needed to pop off the stack for this call, and (b) determine which function is being called.

The dynamic types will have to be checked at some point, either when the value is created (LdFunction/EarlyBind/Deserialization) or called (InvokeFunction).  If we think it will be called at least once, then earlier type-checking would make sense, but if pulling the function data into cache is the slow part of type checking, we might want to do it at Call time.